### ISSUE TITLE
JVM Access Violation in PlatformInternals#internalDeallocateDirectBufferByThunk

### ISSUE THREAD

**Issue by tm-ms** [microstream-one/microstream#1]
With the current release candidate, JM gets an access violation in PlatformInternals#internalDeallocateDirectBufferByThunk when the storage is initialized/loaded.

The used JDK version is identical to the one that code was written with.
All the JDK internals used by that method are present.

If a no-op DirectBufferDeallocator is set to override the behavior, the JVM just hangs at that point, waiting forever for a notification despite no wait/notify or synchronized or any thread controlling code is used.

The PlatformInternals code is written extremely carefully and error-safe:
- offsets to internal fields are retrieved. If something fails there, an internal warning is logged
- the logged warnings are empty
- the method itself checks again every time if viable offsets are present, retrieved instances are not null, are of the assumed type, etc.. If not, nothing is done (as the explicit deallocation is not critical, but rather a fallback-fallback solution for very rare cases)

Nevertheless, there's that access violation and it is currently not in the slightest conceivable what the reason for it is.

**Comment by tm-ms** [comment]
2 reasons:

1.)
The access violation was simply a fogotten null-check. Since the field is accessed via unsafe, a null reference does not just cause a NPE, but crashes the JVM instead.
Funny thing: There's checks and double checks in those methods to cover all kinds of problems or potential future platform incompatibilities. But a simple check for null references (which must be possible), was forgotten...

2.)
The lambda causing the JVM to hang is really strange.
There's no wait/notify or locking or anything similar involved in the MicroStream Code or the passed lambda.
It's a simple method that gets passed a trivial (even empty) lambda.
Still, the thread waits for some object when entering the lambda and never wakes up from that.
Inserting actual code (a syso) did not change that.

What did change the behavior was replacing the lambda with a good old anonymous inner class instance.
With that, it worked normally.

The conclusion from that is that there must be some kind of (extremely rare?) but in the JVM when calling (maybe when on-demand defining) a lambda.

A very similarily sounding issue was already found and described here:
https://stackoverflow.com/questions/30349206/java8-hangs-up-if-getstacktrace-is-called-by-one-thread-and-lambda-definition

That particular situation does not apply here (the JVM argument was not set), but its probably a similar bug or the same bug but caused via a different way.
Maybe it has something to do with additional hot code deployment tooling that is used in the project, because I don't have that installed and I can't reproduce the behavior.

Bottom line:
*1 was easily fixed, of course.
*2 must be checked off as "The JVM can rarely hang on calling a lambda for no apparent reason". ¯\\_(ツ)_/¯

**Comment by tm-ms** [comment]
Tested locally, deployed, test in project. Works.

**Comment by tm-ms** [comment]
**Changes:**
- Fixed a missing null check.
- Can't do anything about JVM crashing on entering a lambda under some mysterious circumstances.

**Migration:**
Nothing required.
